src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/SessionContextValueResolver.php

```diff
@@ -0,0 +1,55 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpKernel\Controller\ArgumentResolver;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Attribute\MapSessionContext;
+use Symfony\Component\HttpKernel\Controller\ValueResolverInterface;
+use Symfony\Component\HttpKernel\ControllerMetadata\ArgumentMetadata;
+
+final class SessionContextValueResolver implements ValueResolverInterface
+{
+    public function resolve(Request $request, ArgumentMetadata $argument): array
+    {
+        if (!$request->hasSession()) {
+            return [];
+        }
+
+        $attribute = $argument->getAttributesOfType(MapSessionContext::class, ArgumentMetadata::IS_INSTANCEOF)[0]
+            ?? null;
+
+        if (!$attribute) {
+            return [];
+        }
+
+        $type = $argument->getType();
+        if (null === $type) {
+            throw new \LogicException(sprintf('#[MapSessionContext] cannot be used on controller argument "$%s" with no type; a class must be provided.', $argument->getName()));
+        }
+        if (!class_exists($type)) {
+            throw new \LogicException(sprintf('#[MapSessionContext] cannot be used on controller argument "$%s" of type "%s"; a class must be provided.', $argument->getName(), $type));
+        }
+
+        if ($request->getSession()->has($type)) {
```

Comment:
Instead of mapping by type, I would map by name.

Reply:
By using the type, we avoid errors related to typo : 
```
__invoke(#[MapSessionContext] BrowsingSessionParameter $browsingSessionParameter)
__invoke(#[MapSessionContext] BrowsingSessionParameter $browsingSessionParameters)
```
In this case, both controler would receive the same object.

We also avoid errors if two different type are used with the same name : 
```
__invoke(#[MapSessionContext] BrowsingSessionParameter $session)
__invoke(#[MapSessionContext] OAuthSessionParameter $session)
```

If it's to keep the session entry name simple, i would rather go with an attribute on the class.
Otherwise, can you specify why you would use the name ?

Reply:
But we prevent having two session parameters sharing the same implementation. This looks like a more serious problem to me.